src/query/set_query.rs

```diff
@@ -0,0 +1,201 @@
+use tantivy_fst::raw::CompiledAddr;
+use tantivy_fst::{Automaton, Map};
+
+use crate::query::{AutomatonWeight, Query, Weight};
+use crate::schema::Field;
+use crate::{Searcher, Term};
+
+/// A Term Set Query matches all of the documents containing any of the Term provided
+///
+/// Terms not using the right Field are discared.
+#[derive(Debug, Clone)]
+pub struct TermSetQuery {
+    field: Field,
+    terms: Vec<Term>,
+}
+
+impl TermSetQuery {
+    /// Create a Term Set Query
+    pub fn new<T: IntoIterator<Item = Term>>(field: Field, terms: T) -> Self {
+        let mut terms: Vec<_> = terms
+            .into_iter()
+            .filter(|key| key.field() == field)
+            .collect();
+        terms.sort_unstable();
+        terms.dedup();
+        TermSetQuery { field, terms }
+    }
+
+    fn specialized_weight(
+        &self,
+        searcher: &Searcher,
+    ) -> crate::Result<AutomatonWeight<SetDfaWrapper>> {
+        let field_entry = searcher.schema().get_field_entry(self.field);
+        let field_type = field_entry.field_type();
+        if !field_type.is_indexed() {
+            let error_msg = format!("Field {:?} is not indexed.", field_entry.name());
+            return Err(crate::TantivyError::SchemaError(error_msg));
+        }
+        let field_type = field_type.value_type();
+
+        // In practice this won't fail because:
+        // - we are writing to memory, so no IoError
+        // - BTreeSet are ordered, and we limit ourselves to values with a fixed prefix (which we
+        // strip), so Map::from_iter get values in order
+        let map = Map::from_iter(
+            self.terms
+                .iter()
+                .filter(|key| key.typ() == field_type)
+                .map(|key| (key.value_bytes(), 0)),
+        )
+        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
+
+        Ok(AutomatonWeight::new(self.field, SetDfaWrapper(map)))
+    }
+}
+
+impl Query for TermSetQuery {
+    fn weight(
+        &self,
+        searcher: &Searcher,
+        _scoring_enabled: bool,
+    ) -> crate::Result<Box<dyn Weight>> {
+        Ok(Box::new(self.specialized_weight(searcher)?))
+    }
+}
+
+struct SetDfaWrapper(Map<Vec<u8>>);
+
+impl Automaton for SetDfaWrapper {
+    type State = Option<CompiledAddr>;
+
+    fn start(&self) -> Self::State {
+        Some(self.0.as_ref().root().addr())
+    }
+    fn is_match(&self, state: &Self::State) -> bool {
+        state
```

Comment:
That one is a personal nitpick but I tend to find the `if-let` style
more readable.

```rust
if let Some(state) = state_opt {
    self.0.as_ref().node(state).is_final()
} else {
    false
}
```